lib/AST/DocComment.cpp

```diff
@@ -334,126 +352,165 @@ swift::extractCommentParts(swift::markup::MarkupContext &MC,
+                            VD->getFullName(), NLOptions::NL_ProtocolMembers,
+                            members);
 
-        SmallVector<const swift::markup::MarkupASTNode *, 8> BodyNodes {
-          Parts.BodyNodes.begin(),
-          Parts.BodyNodes.end()
-        };
-        BodyNodes.push_back(CascadeNote);
-        Parts.BodyNodes = MC.allocateCopy(llvm::makeArrayRef(BodyNodes));
+  for (auto *member : members) {
+    if (!isa<ProtocolDecl>(member->getDeclContext()) ||
+        !member->isProtocolRequirement() ||
+        member->getRawComment().isEmpty())
+      continue;
+    if (requirement)
+      // Found two or more decls with doc-comment.
+      return nullptr;
 
-        return new (MC) DocComment(D, Doc, Parts);
-      }
+    requirement = member;
+  }
+  return requirement;
+}
 
-      BaseDecl = BaseDecl->getOverriddenDecl();
+const ValueDecl *findRequirementDeclWithDocComment(const ValueDecl *VD) {
+  std::queue<const ValueDecl *> requirements;
+  while (true) {
+    for (auto *req : VD->getSatisfiedProtocolRequirements()) {
+      if (!req->getRawComment().isEmpty())
+        return req;
+      else
+        requirements.push(req);
     }
+    if (requirements.empty())
+      return nullptr;
+    VD = requirements.front();
+    requirements.pop();
   }
-  
-  return None;
 }
+} // namespace
 
-static Optional<DocComment *>
-getProtocolRequirementDocComment(swift::markup::MarkupContext &MC,
-                                 const ProtocolDecl *ProtoExt,
-                                 const Decl *D) {
-
-  auto getSingleRequirementWithNonemptyDoc = [](const ProtocolDecl *P,
-                                                const ValueDecl *VD)
-    -> const ValueDecl * {
-      SmallVector<ValueDecl *, 2> Members;
-      P->lookupQualified(const_cast<ProtocolDecl *>(P),
-                         VD->getFullName(),
-                         NLOptions::NL_ProtocolMembers,
-                         Members);
-    SmallVector<const ValueDecl *, 1> ProtocolRequirements;
-    for (auto Member : Members)
-      if (isa<ProtocolDecl>(Member->getDeclContext()) &&
-          Member->isProtocolRequirement())
-        ProtocolRequirements.push_back(Member);
-
-    if (ProtocolRequirements.size() == 1) {
-      auto Requirement = ProtocolRequirements.front();
-      if (!Requirement->getRawComment().isEmpty())
-        return Requirement;
-    }
+const Decl *swift::getDocCommentProvidingDecl(const Decl *D) {
+  if (!D->canHaveComment())
+    return nullptr;
+
+  if (!D->getRawComment().isEmpty())
+    return D;
 
+  auto *VD = dyn_cast<ValueDecl>(D);
+  if (!VD)
     return nullptr;
-  };
 
-  if (const auto *VD = dyn_cast<ValueDecl>(D)) {
-    SmallVector<const ValueDecl *, 4> RequirementsWithDocs;
-    if (auto Requirement = getSingleRequirementWithNonemptyDoc(ProtoExt, VD))
-      RequirementsWithDocs.push_back(Requirement);
+  if (auto *overriden = findOverriddenDeclWithDocComment(VD))
```

Comment:
could we reuse 'getOverriddenDecls()' from indexer.cpp? (see here: https://github.com/apple/swift/blob/master/lib/Index/Index.cpp#L1619). This utility has been used by migrator and indexer for a while.

Reply:
I think it's possible. But for that, we need non-trivial code moving. So let me merge this as is for now.